src/software/jetson_nano/rotary_encoder/RotaryEncoder.py

```diff
@@ -0,0 +1,141 @@
+import Jetson.GPIO as GPIO
+
+"""
+The state of a rotary encoder is dependent on the values read on pin A and pin B.
+The input read on pin A and pin B can either be 0 or 1 and therefore we have four
+possible states. The encoder will transition through these four states during a 
+single rotation and we can use these state transitions to determine whether the encoder 
+has been rotated clockwise or counterclockwise.
+
+Channel B and A go through 4 states for a single rotation.
+The following are values for clockwise rotation:
+  Channel B Channel A
+     0          0
+     0          1
+     1          1
+     1          0
+
+The defined states below are tuples where the first index is the value read from pin B and
+the second index the value read from pin A.
+"""
+STATE_0 = (0, 0)
+STATE_1 = (0, 1)
+STATE_2 = (1, 1)
+STATE_3 = (1, 0)
+
+NUM_STATES = 4
+
+CLOCKWISE = 1
+COUNTERCLOCKWISE = -1
+IDLE = 0
+
+# curr_state : next_state  indicates clockwise rotation
+clockwise_state_transitions = {
+    STATE_0: STATE_1,
+    STATE_1: STATE_2,
+    STATE_2: STATE_3,
+    STATE_3: STATE_0,
+}
+
+# curr_state : next_state  indicates counter clockwise rotation
+counterclockwise_state_transitions = {
+    STATE_0: STATE_3,
+    STATE_1: STATE_0,
+    STATE_2: STATE_1,
+    STATE_3: STATE_2,
+}
+
+"""
+@param A_PIN, the Jetson Nano board pin number used for channel A input of rotary encoder; pin requires pull up resistor
+@param B_PIN, the Jetson Nano board pin number used for channel B input of rotary encoder; pin requires pull up resistor
+@param BUTTON_PIN, the Jetson Nano board pin number used for push button input of rotary encoder; pin requires pull up resistor
+@param on_clockwise_rotate, callback function to be called when rotating clockwise, callback takes no params
+@param on_counterclockwise_rotate, callback function to be called when rotatin counterclockwise, callback takes no params
+@param on_click, callback function to be called when push button is clicked, callback takes no params
+"""
+
+
+class RotaryEncoder:
+    def __init__(
+        self,
+        A_PIN,
+        B_PIN,
+        BUTTON_PIN,
+        on_clockwise_rotate,
+        on_counterclockwise_rotate,
+        on_click,
+    ):
+        self.A_PIN = A_PIN
```

Comment:
I think the arguments for the constructor should be in the function rather than outside the class

Reply:
Hi Jon, I don't quite follow what you mean by this. Do you mean that when someone uses this class they will not pass any aruguments?

Reply:
whoops I mean the _comments_ should be in the function

Reply:
fixed